pkg/controller/logstorage/users/user_controller.go

```diff
@@ -0,0 +1,254 @@
+	reqLogger.Info("Reconciling LogStorage - Users")
+
+	// We skip requests without a namespace specified in multi-tenant setups.
+	if r.multiTenant && request.Namespace == "" {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if this is a tenant-scoped request.
+	tenant, tenantID, err := utils.GetTenant(ctx, r.multiTenant, r.client, request.Namespace)
+	if errors.IsNotFound(err) {
+		reqLogger.Info("No Tenant in this Namespace, skip")
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "An error occurred while querying Tenant", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Get LogStorage resource.
+	logStorage := &operatorv1.LogStorage{}
+	key := utils.DefaultTSEEInstanceKey
+	err = r.client.Get(ctx, key, logStorage)
+	if err != nil {
+		// Not finding the LogStorage CR is not an error, as a Managed cluster will not have this CR available but
+		// there are still "LogStorage" related items that need to be set up
+		if !errors.IsNotFound(err) {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "An error occurred while querying LogStorage", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+		logStorage = nil
+		r.status.OnCRNotFound()
+	}
+
+	// We found the LogStorage instance (and Tenant instance if in multi-tenant mode).
+	r.status.OnCRFound()
+
+	// Wait for the initializing controller to indicate that the LogStorage object is actionable.
+	if logStorage.Status.State != operatorv1.TigeraStatusReady {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Waiting for LogStorage defaulting to occur", nil, reqLogger)
+		return reconcile.Result{}, nil
+	}
+
+	// Wait for Elasticsearch to be installed and available. We don't need to do this in multi-tenant mode because because
+	// we disable kube-controllers ES access in this mode.
+	elasticsearch, err := utils.GetElasticsearch(ctx, r.client)
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "An error occurred trying to retrieve Elasticsearch", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	if elasticsearch == nil || elasticsearch.Status.Phase != esv1.ElasticsearchReadyPhase {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Waiting for Elasticsearch cluster to be operational", nil, reqLogger)
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Query any existing username and password for this Linseed instance. If one already exists, we'll simply
+	// use that. Otherwise, generate a new one.
+	linseedUser := utils.LinseedUser(tenantID)
+	basicCreds := corev1.Secret{}
+	credentialSecrets := []client.Object{}
+	key = types.NamespacedName{Name: render.ElasticsearchLinseedUserSecret, Namespace: helper.TruthNamespace()}
+	if err := r.client.Get(ctx, key, &basicCreds); err != nil && !errors.IsNotFound(err) {
```

Comment:
We have utils.GetSecret() in this repo already that would make this code a bit cleaner.

Reply:
If I had my way, `utils.GetSecret` would be removed and we'd do everything this way... 

I find it an unnecessary abstraction that just intercepts and hides a particular type of error leading to code that is confusingly different from all of the other similar places that we query information from the k8s API. Many many times I have found or written in bugs myself as a result of this helper function, because I expect `errors.IsNotFound` to work, but it doesn't. 

We haven't written that helper function for all of the other types that we might want to query (ConfigMap, ManagementClusterConnection, DaemonSet, etc., etc.,) and of course we shouldn't.

TL;DR - utils.GetSecret shouldn't exist IMO, and I may make a PR to actually eliminate it in the near future assuming I can get people on board with it.

Reply:
I suppose we can agree to disagree about the convenience of the function.

But one immediate question that I have when reading the code the way it is, is why is basicCreds redefined? Isn't that redundant?

Reply:
Yeah, we could probably do without the reassignment. I usually do that sort of things to be safe with regards to side-effects of the calling code, as I don't know what guarantees it provides regarding inputs. In this case, I think we can be pretty confident that it's not modifying fields on `basicCreds` in the case of an error, though.